### ISSUE TITLE
String type char * vs char const *

### ISSUE THREAD

**Issue by Matthew-Jemielity** [protobuf-c/protobuf-c#239]
Current protobuf-c (commit 006d69bd8481738323c58af5cd4c8477d09eb9f5) translates string type into char _:
`
$ cat msg.proto
syntax = "proto2";
message m {
  required string s = 1;
}
$ protoc-c msg.proto --c_out=.  
$ cat msg.pb-c.h
/_ Generated by the protocol buffer compiler.  DO NOT EDIT! */
[...]
struct  _M
{
  ProtobufCMessage base;
  char *s;
};
[...]
`
Does any code depend on that translated type being non-const or changes contents of the string? If not, then in my opinion it would be better to translate string into char const *. Constness would allow me to, for example, assing string literals to M.s without complaints from the compiler.

**Comment by jin-eld** [comment]
@lipnitsk hi, is there any ETA on a new release? :) I am looking forward to the const_strings option, I found it on the `next` branch and I was wondering if it's considered stable stable enough? Then I'd simply go with it until the new release is out. Thanks!

**Comment by lipnitsk** [comment]
I pinged @edmonds to see if he can cut a release soon. Take a look at the commit in this PR and the test-full.proto change to see the example of how to activate this option.

**Comment by jin-eld** [comment]
> I pinged @edmonds to see if he can cut a release soon. 

Cool, thank you!

> Take a look at [dfb6e30#diff-89aab21e01477310e00342624684980d34c2ce8a2e0f2950a147519ad52d4ccbR5](https://github.com/protobuf-c/protobuf-c/commit/dfb6e300fab0aec0093eeed4e7416795091f6e5f#diff-89aab21e01477310e00342624684980d34c2ce8a2e0f2950a147519ad52d4ccbR5) to see the example of how to activate this option.

Well, this is what I have been looking at, but I seem to be missing something.

I understand this is not meant literally: `option (pb_c_file).const_strings = true;` ? Actually I tried it like that too and it said `Option "(pb_c_file)" unknown. Ensure that your proto definition file imports the proto which defines the option.`

So I then tried using `foo.pb-c.c` and alike, but it then was only complaining `Expected ")"`

What exactly is `pb_c_file` or what is supposed to go into the parens?

**Comment by lipnitsk** [comment]
Make sure your proto file imports `protobuf-c/protobuf-c.proto`, just like the error says.

**Comment by jin-eld** [comment]
> Make sure your proto file imports `protobuf-c/protobuf-c.proto`, just like the error says.

Oooh... hmm,, I want to make sure I understand this whole thing correctly: let's assume there are different projects in different languages that are using the defined protocol. Does this mean that they can't share the same `.proto` definition, unless the others who are not using `protobuf-c` also download/include the `protobuf-c.proto` file or is it possible to make the import optional somehow?

I would have expected this option to be a command line parameter to the proto-c compiler or something like that since this is purely a compiler/generator setting... Am I supposed to pull `protobuf-c.proto` into my project and distribute it along with my sources, to make sure that whoever wants to implement the protocol supported by my application on anything but C can just take and compile my `.proto` files? What is the suggested way of handling this?

**Comment by lipnitsk** [comment]
Yeah, it's not ideal, but adding command line options doesn't scale well - see discussion here if interested: https://github.com/protobuf-c/protobuf-c/issues/204

**Comment by jin-eld** [comment]
Oh I see similar arguments have been exchanged there already. Apart from the issues that @pbor mentioned, as in - not wanting to have compiler specific options in shared .proto files, I also see the following issue:

it gets quite messy in regard to the -I include directory parameters that need to be passed to `proto-c`:  since `protobuf-c.proto` is including `google/protobuf/descriptor.proto` I now have to also depend on the `protobuf` package and check where it has installed its `descriptor.proto`, which I normally should not be worried about, since it's only a dependency of `protobuf-c`.

So while I kind of do understand your reasoning regarding the command line, putting compiler specific settings in the actual `.proto` description works against the whole idea that made protocol buffers an interesting choice: having a shared `.proto` definition that would work for any language binding and any compiler that supports protocol buffers. Putting settings of one particular compiler in there just does not feel right... so again I agree with @pbor here, that compiler specific settings belong into the build. The current implementation would basically force me to patch protobuf-c in order to keep the shared `.proto` files "clean".

Could we move those options out to a separate file that would get fed to the compiler together with the actual .proto that is being compiled? This could be just one additional command line, like `--compiler-config myoptions.proto` or something like that? Main goal would be to separate the protocol definition (which usually gets shared between various projects) and compiler specific settings.  Avoiding to know the location of `google/protobuf/descriptor.proto` would also be very helpful. 

Basically, from user perspective, I do not care that much how exactly those options can be set as long as they:
* do not "pollute" the shared protocol definition `.proto` files
* do not require the application to dig into dependency chains/locations to make sure that nested `import` statements that are internal to the library do not fail

Does that sound reasonable?

**Comment by lipnitsk** [comment]
Yeah, unfortunately we rely on the upstream `protoc` code to do the .proto parsing, so I'm not sure there is a way to inject imports, but if there is a way it would be great!

**Comment by jin-eld** [comment]
OK, maybe let's take a step back, I'm trying to understand the whole picture :) Does it have to be solved via proto definitions, or could the proto-c compiler config file be something different? My understanding was that you chose the `.proto`  format simply because the parser was already there? Or is the upstream proto compiler using the same options mechanism, by having them directly in the `.proto` file and is that the reason for `proto-c` taking the same route?

While you did not like the command line option variant - would it be acceptable to simply have both? If someone does not mind having compiler options in their protocols - fine, those who need to share the `.proto` files could still keep them clean and use command line options. Just think of `gcc` and the tons of options it has - it still works and we all use it...
